_posts/2024-10-31-WildFly-playing-with-generative-ai.adoc

```diff
@@ -0,0 +1,355 @@
+It helps by configuring and providing all the required elements for such an application.
+
+=== What is RAG ?
+
+As its name indicates, RAG is about retrieving relevant elements from pieces of information from your data and using them to enrich the prompt before sending the whole query (initial prompt + relevant data) to the LLM. 
+The goal is to provide sufficient context to the LLM so that it can use those pieces of information and not hallucinate.
+
+====  The core elements of RAG
+
+As you can see on the schema below, you need a few elements to provide a RAG functionality:
+
+* some components to get relevant pieces of information. We will focus on using a 'semantic' search approach, a.k.a. vector search, so we will need:
+    - an embedding model to create embeddings (aka semantic vectors) from the user query thus creating an embedding query.
+    - an embedding store containing all the embeddings and the associated segments of data and against which/ the embedding query will be run.
+* a client to the LLM to send the enriched query against.
+
+image::ai/rag.png[RAG]
+
+==== RAG with LangChain4j
+
+The current feature pack provides a basic integration with the library LangChain4J. That means that currently we are exposing the LangChain4J API to the application.
+So it provides:
+
+* `dev.langchain4j.model.embedding.EmbeddingModel`
+* `dev.langchain4j.store.embedding.EmbeddingStore````
+* ``dev.langchain4j.rag.content.retriever.ContentRetriever````
+* `dev.langchain4j.model.chat.ChatLanguageModel```
+
+So you can configure instances of these objects via the WildFly management API to be injected via CDI into your application.
+
+We also support partially LangChain4J https://docs.langchain4j.dev/tutorials/ai-services/[*AI Services*, window=_blank] using https://github.com/smallrye/smallrye-llm/blob/0.0.1/smallrye-llm-langchain4j-core/src/main/java/io/smallrye/llm/spi/RegisterAIService.java[RegisterAIService, window=_blank] from https://github.com/smallrye/smallrye-llm[smallrye-llm, window=_blank]. Please note that this will change in the future as this project is quite new.
+
+== The feature pack
+
+The feature pack is composed of a subsystem and a CDI extension. It supports several layers based on what you are trying to acheive. The list is not exhaustive and may vary in the future.
+
+==== The layers
+
+This feature pack provides several layers to choose from when provisioning your server:
+
+* Support for chat models to interact with a LLM via `ChatModelLanguage`:
+** `mistral-ai-chat-model`: provides integration with https://mistral.ai/[Mistral AI, window=_blank]
+** `ollama-chat-model`: provides integration with https://ollama.com/[Ollama, window=_blank]
+** `openai-chat-model`: provides integration with https://openai.com/[OpenAI, window=_blank] (current configuration target https://console.groq.com/playground[Groq])
+* Support for embedding models via `EmbeddingModel`: 
+  * `in-memory-embedding-model-all-minilm-l6-v2`
```

Comment:
These entries don't show as indented. See https://github.com/wildfly/wildfly.org/blob/f6a6ad40dd79e59af7cd252d100ad172a419d390/_posts/2024-10-31-WildFly-playing-with-generative-ai.adoc

Also, a brief explanation of what these things are might be good